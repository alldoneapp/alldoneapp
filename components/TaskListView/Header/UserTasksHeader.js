import React, { useState } from 'react'
import { StyleSheet, Text, TouchableOpacity, View } from 'react-native'
import PropTypes from 'prop-types'
import Popover from 'react-tiny-popover'
import { useSelector, useDispatch } from 'react-redux'

import styles from '../../styles/global'
import TaskHeaderMoreButton from '../../UIComponents/FloatModals/MorePopupsOfMainViews/Tasks/TaskHeaderMoreButton'
import ChangeObjectListModal from '../../UIComponents/FloatModals/ChangeObjectListModal'
import { translate } from '../../../i18n/TranslationService'
import ToggleByTime from '../ToggleByTime'
import { checkIfSelectedAllProjects } from '../../SettingsView/ProjectsSettings/ProjectHelper'
import { switchProject, storeLoggedUser } from '../../../redux/actions'
import store from '../../../redux/store'

const UserTasksHeader = () => {
    const dispatch = useDispatch()
    const smallScreenNavigation = useSelector(state => state.smallScreenNavigation)
    const isMiddleScreen = useSelector(state => state.isMiddleScreen)
    const userId = useSelector(state => state.currentUser.uid)
    const currentSection = useSelector(state => state.taskViewToggleSection)
    const selectedProjectIndex = useSelector(state => state.selectedProjectIndex)
    const project = useSelector(state => state.loggedUserProjects[selectedProjectIndex])
    const realProjectIds = useSelector(state => state.loggedUser.realProjectIds)
    const isAnonymous = useSelector(state => state.loggedUser.isAnonymous)
    const [open, setOpen] = useState(false)

    const inAllProjects = checkIfSelectedAllProjects(selectedProjectIndex)
    const inOpenSection = currentSection === 'Open'

    const accessGranted = !isAnonymous && (inAllProjects || (project && realProjectIds.includes(project.id)))

    const handleToggle = showByTime => {
        if (showByTime && !inAllProjects) {
            // First update the showAllProjectsByTime setting and switch to All Projects view
            const { loggedUser } = store.getState()
            dispatch([storeLoggedUser({ ...loggedUser, showAllProjectsByTime: true }), switchProject(-1)])
        }
    }

    return (
        <View
            style={[
                localStyles.container,
                smallScreenNavigation
                    ? localStyles.headerTextForMobile
                    : isMiddleScreen && localStyles.headerTextForTablet,
            ]}
        >
            <Popover
                content={<ChangeObjectListModal closePopover={() => setOpen(false)} />}
                onClickOutside={() => setOpen(false)}
                isOpen={open}
                position={['bottom', 'left', 'right', 'top']}
                padding={4}
                align={'start'}
                contentLocation={smallScreenNavigation ? null : undefined}
            >
                <TouchableOpacity disabled={!accessGranted} accessible={false} onPress={() => setOpen(true)}>
                    <Text style={localStyles.headerText}> {translate('Tasks')}</Text>
                </TouchableOpacity>
            </Popover>
            {inOpenSection && <TaskHeaderMoreButton userId={userId} />}
            <ToggleByTime onToggle={handleToggle} containerStyle={!inOpenSection && { marginLeft: 8 }} />
        </View>
    )
}

UserTasksHeader.propTypes = {
    style: PropTypes.any,
}

export function getFormattedName(fullName) {
    if (fullName === 'Loading...') {
        return fullName
    }

    let name = fullName.split(' ')[0]

    if (name[name.length - 1] === 's') {
        return name + `' tasks`
    }

    return name + `'s tasks`
}

const localStyles = StyleSheet.create({
    container: {
        flex: 1,
        height: 80,
        maxHeight: 80,
        minHeight: 80,
        flexDirection: 'row',
        alignItems: 'center',
        paddingTop: 40,
        paddingBottom: 8,
        paddingHorizontal: 104,
    },
    headerText: {
        ...styles.title5,
    },
    headerTextForMobile: {
        paddingHorizontal: 16,
    },
    headerTextForTablet: {
        paddingHorizontal: 56,
    },
})

export default UserTasksHeader
