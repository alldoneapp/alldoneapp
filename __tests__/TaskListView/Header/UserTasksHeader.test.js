import React from 'react'
import UserTasksHeader, { getFormattedName } from '../../../components/TaskListView/Header/UserTasksHeader'

import renderer from 'react-test-renderer'

describe('UserTasksHeader component', () => {
    describe('UserTasksHeader empty snapshot test', () => {
        it('should render correctly for basic header', () => {
            const tree = renderer.create(<UserTasksHeader headerText="Martina Muller" isName={true} />).toJSON()
            expect(tree).toMatchSnapshot()
        })
        it('should render correctly for header with caption', () => {
            const tree = renderer
                .create(<UserTasksHeader headerText="Done project task" headerCaption={'5 Open'} />)
                .toJSON()
            expect(tree).toMatchSnapshot()
        })
    })

    describe('getFormattedName function test', () => {
        it('should correctly format a given name to display at the tasks view when the name does not end with s', () => {
            const name = getFormattedName('Martina')
            expect(name).toBe(`Martina's tasks`)
        })

        it('should correctly format a given name to display at the tasks view when the name ends with s', () => {
            const name = getFormattedName('Charles')
            expect(name).toBe(`Charles' tasks`)
        })
    })
})
