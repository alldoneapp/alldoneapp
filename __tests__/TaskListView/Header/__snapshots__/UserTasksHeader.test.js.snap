// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserTasksHeader component UserTasksHeader empty snapshot test should render correctly for basic header 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
        "height": 80,
        "maxHeight": 80,
        "minHeight": 80,
        "paddingBottom": 8,
        "paddingTop": 40,
      },
      undefined,
    ]
  }
>
  <Text
    style={
      Object {
        "fontFamily": "Roboto-Regular",
        "fontSize": 24,
        "lineHeight": 32,
        "marginRight": 14,
      }
    }
  >
    Martina's tasks
  </Text>
  <View
    style={
      Object {
        "height": 28,
        "justifyContent": "flex-end",
        "marginLeft": 16,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#04142F",
            "fontFamily": "Roboto-Regular",
            "fontSize": 12,
            "letterSpacing": 0.48,
            "lineHeight": 20,
          },
          Object {
            "color": "#4E5D78",
          },
        ]
      }
    >
      
    </Text>
  </View>
</View>
`;

exports[`UserTasksHeader component UserTasksHeader empty snapshot test should render correctly for header with caption 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
        "height": 80,
        "maxHeight": 80,
        "minHeight": 80,
        "paddingBottom": 8,
        "paddingTop": 40,
      },
      undefined,
    ]
  }
>
  <Text
    style={
      Object {
        "fontFamily": "Roboto-Regular",
        "fontSize": 24,
        "lineHeight": 32,
        "marginRight": 14,
      }
    }
  >
    Done team task
  </Text>
  <View
    style={
      Object {
        "height": 28,
        "justifyContent": "flex-end",
        "marginLeft": 16,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "#04142F",
            "fontFamily": "Roboto-Regular",
            "fontSize": 12,
            "letterSpacing": 0.48,
            "lineHeight": 20,
          },
          Object {
            "color": "#4E5D78",
          },
        ]
      }
    >
      5 Open
    </Text>
  </View>
</View>
`;
